### PR TITLE
Add extensive 'verbose', 'debug', and 'silly' logging

### LINKED ISSUE #635: Add extensive logging for better debugging of Dredd
Currently, `--level=verbose` or `--level=silly` have basically zero effect on Dredd and do not help to debug any Dredd-related problems at all. The lowest level of messages Dredd now produces is more or less `info`, which contains informative messages for Dredd user, not debugging messages helpful for resolving issues with Dredd itself.

This requirement is driven by my own personal need when developing Dredd as well as by resolving issues of actual Dredd users (@philsturgeon, or @zmorris in #633). Related also to #625.

### PR DESCRIPTION
**Pull request by honzajavorek**
### What is this change about?

- Primarily, this PR closes #635. Please see description in the issue.
- It also prints some debugging info (Dredd version, system version, node version, ...) in the beginning, which is a prototype of https://github.com/apiaryio/dredd/issues/625. We agreed upon this with @netmilk - better ugly and now than beautiful and later.
- During works on this PR I also found out my computer is prone to https://github.com/nodejs/node/issues/6456. The Node.js issue mentions a workaround. I added this workaround to `bin/dredd` to avoid the issue for myself and any other users who might have been affected.

### Why test coverage dropped?

I believe it's because I added a lot of new lines, also to places which were uncovered previously. That made the uncovered parts statistically larger than they were.

### CODE REVIEW

File: src/add-hooks.coffee [hunk 1]
```diff
@@ -44,11 +44,12 @@ addHooks = (runner, transactions, callback) ->
...
 
     catch error
-      logger.warn 'Skipping hook loading...'
-      logger.warn 'Error reading hook file "' + filePath + '"'
-      logger.warn 'This probably means one or more of your hookfiles is invalid.'
-      logger.warn 'Message: ' + error.message if error.message?
-      logger.warn 'Stack: ' + error.stack if error.stack?
+      logger.warn("""\
+        Skipping hook loading. Error reading hook file '#{filePath}'. \
+        This probably means one or more of your hook files is invalid.
```
Comment (miiila) [inline src/add-hooks.coffee]: NIT: `s/is/are`

File: src/add-hooks.coffee [hunk 2]
```diff
@@ -87,9 +88,9 @@ addHooks = (runner, transactions, callback) ->
         loadSandboxHooksFromStrings(callback)
       else
         # not sandboxed code can't be loaded from the string
-        msg = """
-        Not sandboxed hooks loading from strings is not implemented,
-        Sandbox mode must be enabled when loading hooks from strings."
+        msg = """\
```
Comment (miiila) [inline src/add-hooks.coffee]: NIT: Unnecessary `"""`.

File: src/dredd-command.coffee [hunk 3]
```diff
@@ -9,13 +9,15 @@ spawnArgs = require 'spawn-args'
 Dredd = require './dredd'
 interactiveConfig = require './interactive-config'
 configUtils = require './config-utils'
-logger = require './logger'
+logger = require('./logger')
 
 packageData = require('../package.json')
 
+
```
Comment (miiila) [inline src/dredd-command.coffee]: Is the new line necessary? Same for line 20.

File: src/dredd-command.coffee [hunk 4]
```diff
@@ -54,17 +56,19 @@ class DreddCommand
...
+      return callback()
 
     term = =>
-      logger.info 'Sending SIGTERM to the backend server'
+      logger.info('Sending SIGTERM to backend server process.')
       @serverProcess.kill 'SIGTERM'
 
     kill = =>
-      logger.info 'Killing backend server'
+      logger.info('Killing backend server process.')
```
Comment (miiila) [inline src/dredd-command.coffee]: I'd keep `'Sending SIGKILL...` wording, but YMMV.

File: src/hooks-worker-client.coffee [hunk 5]
```diff
@@ -29,14 +29,19 @@ class HooksWorkerClient
     @emitter = new EventEmitter
 
   start: (callback) ->
+    logger.verbose('Looking up hooks handler implementation:', @language)
     @setCommandAndCheckForExecutables (executablesError) =>
```
Comment (miiila) [inline src/hooks-worker-client.coffee]: Unnecessary fat arrow.  Same applies to similar statements on lines below.

File: src/reporters/apiary-reporter.coffee [hunk 6]
```diff
@@ -177,38 +181,34 @@ class ApiaryReporter
...
 
-      res.on 'end', =>
-        if @verbose
-          logger.log 'Rest Reporter Response ended'
+      res.on 'end', ->
+        logger.verbose('Apiary reporter Response received.')
 
         try
           parsedBody = JSON.parse buffer
         catch e
```
Comment (miiila) [inline src/reporters/apiary-reporter.coffee]: Is caught error ignored intentionally? Wouldn't it help for faster orientation in JSON parse error?

File: src/reporters/apiary-reporter.coffee [hunk 7]
```diff
@@ -222,18 +222,14 @@ class ApiaryReporter
...
 
     unless @configuration['apiToken'] == null
       options.headers['Authentication'] = 'Token ' + @configuration['apiToken']
 
-    if @verbose
-      info =
-        options: options
-        body: body
-      logger.log 'Rest Reporter Request:', JSON.stringify(info, null, 2)
+    logger.verbose('Apiary reporter request:', JSON.stringify({options: options, body}, null, 2))
```
Comment (miiila) [inline src/reporters/apiary-reporter.coffee]: Looks weird with one extra key only.

File: src/dredd-command.coffee (hunk 3 above)
Comment (honzajavorek) [inline src/dredd-command.coffee]: I usually use two lines to separate top-level module content and one line to separate everything else. The CS style guide surprised me by suggesting that after `class` there should be three new lines or something like that, so if that popped up on me, I obeyed, although I don't think 3 lines are in any way better than 2.

File: src/dredd-command.coffee (hunk 4 above)
Comment (honzajavorek) [inline src/dredd-command.coffee]: I'm afraid many developers and users of Dredd do not know anything about signals, so I'm trying to move away from these terms, at least for user-focused `info` logging (I'm fine with such expressions in `debug` etc.). Also, I'm not sure SIGKILL will be appropriate term to use when we'll start supporting Windows, would need to check.

File: src/hooks-worker-client.coffee (hunk 5 above)
Comment (honzajavorek) [inline src/hooks-worker-client.coffee]: Yes. But I didn't want to snowball even more than I already did 😞 It would be probably worth to make a dedicated PR to fix some of the most outstanding linter issues in Dredd codebase.

File: src/reporters/apiary-reporter.coffee (hunk 7 above)
Comment (honzajavorek) [inline src/reporters/apiary-reporter.coffee]: I don't understand your comment. But when I look at the change, `{options: options, body}` could be written as `{options, body}`.

File: src/reporters/apiary-reporter.coffee (hunk 6 above)
Comment (honzajavorek) [inline src/reporters/apiary-reporter.coffee]: Good point! This could be handled in a better way. I'll address this in this PR when you finish the initial review.

File: src/reporters/apiary-reporter.coffee [hunk 8]
```diff
@@ -242,21 +238,18 @@ class ApiaryReporter
       else
         return callback error, req, null
 
-    if @configuration.apiUrl?.indexOf('https') is 0
-      if @verbose
-        logger.log 'Starting REST Reporter HTTPS Request'
+    logger.verbose('Starting Apiary reporter HTTP request')
+    if @configuration.apiUrl.match(/^https/)
```
Comment (miiila) [inline src/reporters/apiary-reporter.coffee]: When compared with previous version `?` is missing. Please consider, whether it should be present or not.

File: src/reporters/apiary-reporter.coffee [hunk 9]
```diff
@@ -242,21 +238,18 @@ class ApiaryReporter
       else
         return callback error, req, null
 
-    if @configuration.apiUrl?.indexOf('https') is 0
-      if @verbose
-        logger.log 'Starting REST Reporter HTTPS Request'
+    logger.verbose('Starting Apiary reporter HTTP request')
```
Comment (miiila) [inline src/reporters/apiary-reporter.coffee]: Sentence `Starting Apiary reporter HTTP request` doesn't make sense for me, but I can be missing some context.

File: src/reporters/apiary-reporter.coffee [hunk 10]
```diff
@@ -242,21 +238,18 @@ class ApiaryReporter
...
-
     else
-      if @verbose
-        logger.log 'Starting REST Reporter HTTP Response'
+      logger.debug('Using unsecured HTTP for Apiary reporter request')
       req = http.request options, handleResponse
 
-      req.on 'error', handleReqError
-
+    req.on 'error', handleReqError
```
Comment (miiila) [inline src/reporters/apiary-reporter.coffee]: `req.on('error', handleReqError)`

File: src/apply-configuration.coffee [hunk 11]
```diff
@@ -13,8 +14,23 @@ coerceToArray = (value) ->
   else
     return value
 
-applyConfiguration = (config) ->
 
+applyLoggingOptions = (options) ->
+  # coerce color to bool
+  if options.color == 'false'
+    options.color = false
+  else if options.color == 'true'
```
Comment (miiila) [inline src/apply-configuration.coffee]: Only strings have special handling here. I would change comment in a way to emphasise it (e.g. `coerce string color to bool, leave others`).

File: src/dredd-command.coffee [hunk 12]
```diff
@@ -8,6 +8,7 @@ spawnArgs = require 'spawn-args'
 
 Dredd = require './dredd'
 interactiveConfig = require './interactive-config'
+{applyLoggingOptions} = require './apply-configuration'
```
Comment (miiila) [inline src/dredd-command.coffee]: I am not sure with the filename `apply-configuration`, when it expose more then one function. Maybe `configuration` only would be enough?

File: src/dredd-command.coffee [hunk 13]
```diff
@@ -266,7 +269,6 @@ class DreddCommand
       , waitMilis
 
   run: ->
-    logger.verbose('Preparing for start of testing.')
```
Comment (miiila) [inline src/dredd-command.coffee]: I am not sure with logging policy, so just pointing this removal for reconsideration, whether it should be really removed or not.

File: bin/dredd [hunk 14]
```diff
@@ -2,6 +2,14 @@
 
 var fs = require('fs');
 
+
+// Ignore this block, it's pure magic, temporary fix
```
Comment (miiila) [inline bin/dredd]: 🎆

File: src/dredd-command.coffee [hunk 15]
```diff
@@ -281,7 +296,8 @@ class DreddCommand
       return if @finished
 
     configurationForDredd = @initConfig()
-    @dreddInstance = @initDredd configurationForDredd
+    @logDebuggingInfo(configurationForDredd)
```
Comment (miiila) [inline src/dredd-command.coffee]: Is it intentional to output the debugging info always? It seems to me with every Dredd run, debug info will be printed.

File: src/apply-configuration.coffee [hunk 16]
```diff
@@ -91,8 +91,8 @@ applyConfiguration = (config) ->
   if config.blueprintPath
     configuration.options.path.push config.blueprintPath
 
-  configuration.options.method.map((method) ->
-    method.toUpperCase()
+  configuration.options.method = (
```
Comment (miiila) [inline src/apply-configuration.coffee]: Shouldn't it be `configuration.options.methods`? I think name should reflect the fact it's array.

File: src/dredd-command.coffee (hunk 13 above)
Comment (honzajavorek) [inline src/dredd-command.coffee]: This logging happens before logging level is set, so it can never make it to the output.

File: src/dredd-command.coffee (hunk 15 above)
Comment (honzajavorek) [inline src/dredd-command.coffee]: Yes. It's poor man's implementation of #625 - with @netmilk we decided we'll just print this kind of info in the beginning of every Dredd run for now.

File: src/apply-configuration.coffee (hunk 16 above)
Comment (honzajavorek) [inline src/apply-configuration.coffee]: The thing is, the Dredd option is `--method`, but accepts also multiple values, so in configuration handling, it's coerced to array while the original singular name sticks. Not ideal and it's on multiple places in Dredd currently. Confused me too, at the first sight.

File: src/hooks-worker-client.coffee (hunk 5 above)
Comment (miiila) [inline src/hooks-worker-client.coffee]: Ah, I missed the indentation, sorry for that.

**Comment by miiila** [comment]
@honzajavorek Could you please just briefly comment https://github.com/apiaryio/dredd/pull/636#discussion-diff-82753444R22 and https://github.com/apiaryio/dredd/pull/636#discussion-diff-82754205R11? It's neither addressed in last commit nor commented, so I don't know, if you missed it or didn't consider as important. Thanks.

**Comment by honzajavorek** [comment]
@miiila Addressed everything (including those two you mentioned) and tests passed.

**Comment by miiila** [comment]
@honzajavorek so LGTM!
